src/ImageManipulation.php

```diff
@@ -782,6 +789,41 @@ public function getHeight()
         return 0;
     }
 
+    /**
+     * Return whether image will be lazy loaded
+     *
+     * @return bool
+     */
+    public function getIsLazyLoaded() : bool
+    {
+        if (Image::getLazyLoadingEnabled() && $this->getWidth() && $this->getHeight()) {
+            return $this->lazyLoad;
+        }
+        return false;
+    }
+
+    /**
+     * Set whether image will be lazy loaded
+     *
+     * @param bool $lazyLoad
+     * @return self $this
+     */
+    public function setLazyLoad(bool $lazyLoad): self
```

Comment:
If you have this in your template
```
<h2>Not Lazy Loaded</h2>
$ScaleWidth(100).setLazyLoad(0)
$setLazyLoad(0)
$setLazyLoad(1).$setLazyLoad(0)

<h2>Lazy Loaded</h2>
$ScaleWidth(100)
$Me
$setLazyLoad(1)
```

All the rendered images will not have a `loading` attribute because only the first rendered image is honoured. If you reverse the two blocks all the Image will have the loading attribute.

`setLazyLoad` needs to return a copy of itself like the other `ImageManipulation` function.

Reply:
This method probably should be start with an upper case letter to match the other image manipulation methods.

Reply:
We need to remove the explicit `bool` type so people can provide the value as `'false'` or `'0'` because of a limitation in our SS rendering engine.

Reply:
> This method probably should be start with an upper case letter to match the other image manipulation methods.

Agree, I figured we should do that as well.

> We need to remove the explicit bool type so people can provide the value as 'false' or '0' because of a limitation in our SS rendering engine.

That's not gonna work because it will just set the variable to `false` as a string.
We should really just expect a 0/1 value like every other template API already does.

Reply:
> All the rendered images will not have a loading attribute because only the first rendered image is honoured. If you reverse the two blocks all the Image will have the loading attribute.
> 
> setLazyLoad needs to return a copy of itself like the other ImageManipulation function.

I'm struggling to see how this would matter in a real world scenario. What would be the point of lazy loading and not lazy load the same image? After all, it's pointing to the same URL.

The only use case I see is if you manipulate the image, and if you do that it will create a copy and toggling the lazy load for the same image with a different variant will work as expected.

Reply:
As a developer, I shouldn't have to worry about how `SetLazyLoad(false)` will be cast internally. It should just work. So in the method, we should just convert a 'false' string to an actual boolean.

> I'm struggling to see how this would matter in a real world scenario. What would be the point of lazy loading and not lazy load the same image? After all, it's pointing to the same URL.

SSViewer tries to cache things as much as possible as it goes along. You could end up with weird scenario where you have an include that gets call in different context. We'll just end up with weird bugs and have inconsistent result if we do it this way.

That's why `ImageManipulation` has all this convoluted logic and jumps through all these hoops to create duplicate object of Image.

Reply:
> As a developer, I shouldn't have to worry about how SetLazyLoad(false) will be cast internally. It should just work. 

Yes, it should. However this is something that should be solved for templates in general, not just hardcoded into individual methods.

It is clear that this is something that will keep being an issue unless we fix it properly, and I don't think this "duct-taping" fix approach is the right way to solve it. https://github.com/silverstripe/silverstripe-cms/commit/ed0680a26464dbc37ef35a3abd4e63014e361fd3

> SSViewer tries to cache things as much as possible as it goes along. You could end up with weird scenario where you have an include that gets call in different context. We'll just end up with weird bugs and have inconsistent result if we do it this way.
> 
> That's why ImageManipulation has all this convoluted logic and jumps through all these hoops to create duplicate object of Image.

I'm not sure about that yet. Do you have an example? I'm under the impression it duplicates the image after altering it / when the variant changes.

Reply:
> > As a developer, I shouldn't have to worry about how SetLazyLoad(false) will be cast internally. It should just work.
> 
> Yes, it should. However this is something that should be solved for templates in general, not just hardcoded into individual methods.
> 
> It is clear that this is something that will keep being an issue unless we fix it properly, and I don't think this "duct-taping" fix approach is the right way to solve it.

FYI this has been resolved in 5.x/dev-master with https://github.com/silverstripe/silverstripe-framework/pull/8152 (and https://github.com/silverstripe/silverstripe-framework/pull/8456/ for null values, which was a marginally more invasive change). I treated it as an API change at the time, but I suppose we could look at whether it’d be possible to cherry-pick those changes into 4.x?

I wouldn’t be opposed to “duct-taping” something like this for now. It’s one line of zero-maintenance code to reduce a bunch of confusion/extra documentation for developers using this.

Reply:
Yeah, I was suggest cherry-picking as well. Might be fine as long as we consider it a bug.

Reply:
It duplicates the PHP object, but it doesn't dupilcate the actual File database entry or the physical file.

The idea is to stack up all the image manipulation calls, but wait until you actually render the image at the end to execute them. Otherwise, you would be running a bunch of expensive (and potentially unneeded) image manipulation operations immediately when calrling a method like `CropHeight`.

My guess is you probably just need to do something like this to duplicate you current object.
```php
$this->manipulate('', function($callback, $store, $filename, $hash, $variant) {
    return [
    'Filename' => $filename,
    'Hash' => $hash,
    'Variant' => $this->getVariant()
]
});
```

Then you set a lazy loading flag on you duplicate.

You need to make sure that your lazy loading flag gets copied across to any follow up calls to `ImageManipulation::manipulate()`.